Plutarch/TryFrom.hs

```diff
@@ -0,0 +1,367 @@
+  ) =>
+  PTryFrom (PBuiltinList PData) (PDataRecord ((name ':= pty) ': as))
+  where
+  type
+    PTryFromExcess (PBuiltinList PData) (PDataRecord ((name ':= pty) ': as)) =
+      HRecP
+        ( '(name, ExcessForField pty)
+            ': UnHRecP (PTryFromExcess (PBuiltinList PData) (PDataRecord as))
+        )
+  ptryFrom' opq = runTermCont $ do
+    h <- tcont $ plet $ phead # opq
+    hv <- tcont $ ptryFrom @(PAsData pty) @PData h
+    t <- tcont $ plet $ ptail # opq
+    tv <- tcont $ ptryFrom @(PDataRecord as) @(PBuiltinList PData) t
+    pure (punsafeCoerce opq, HRecGeneric (HCons (Labeled hv) (coerce $ snd tv)))
+
+newtype Helper a b s = Helper (a s, b s)
+
+instance (Reducible (a s), Reducible (b s)) => Reducible (Helper a b s) where
+  type Reduce (Helper a b s) = (Reduce (a s), Reduce (b s))
+
+instance
+  ( PTryFrom (PBuiltinList PData) (PDataRecord as)
+  , PTryFromExcess (PBuiltinList PData) (PDataRecord as) ~ HRecP ase
+  ) =>
+  PTryFrom PData (PAsData (PDataRecord as))
+  where
+  type
+    PTryFromExcess PData (PAsData (PDataRecord as)) =
+      Helper (Flip Term (PDataRecord as)) (PTryFromExcess (PBuiltinList PData) (PDataRecord as))
+  ptryFrom' opq = runTermCont $ do
+    l <- snd <$> (tcont $ ptryFrom @(PAsData (PBuiltinList PData)) opq)
+    r <- tcont $ ptryFrom @(PDataRecord as) l
+    pure (punsafeCoerce opq, r)
+
+instance {-# OVERLAPPING #-} SumValidation 0 ys => PTryFrom PData (PAsData (PDataSum ys)) where
+  type PTryFromExcess PData (PAsData (PDataSum ys)) = Const ()
+  ptryFrom' opq = runTermCont $ do
+    x <- tcont $ plet $ pasConstr # opq
+    constr <- tcont $ plet $ pfstBuiltin # x
+    fields <- tcont $ plet $ psndBuiltin # x
+    _ <- tcont $ plet $ validateSum @0 @ys constr fields
+    pure (punsafeCoerce opq, ())
+
+class SumValidation (n :: Nat) (sum :: [[PLabeledType]]) where
+  validateSum :: Term s PInteger -> Term s (PBuiltinList PData) -> Term s POpaque
+
+instance
+  {-# OVERLAPPABLE #-}
+  forall (n :: Nat) (x :: [PLabeledType]) (xs :: [[PLabeledType]]).
+  ( PTryFrom (PBuiltinList PData) (PDataRecord x)
+  , SumValidation (n + 1) xs
+  , KnownNat n
+  ) =>
+  SumValidation n (x ': xs)
+  where
+  validateSum constr fields =
+    pif
+      (fromInteger (natVal $ Proxy @n) #== constr)
+      ( unTermCont $ do
+          _ <- tcont $ ptryFrom @(PDataRecord x) fields
+          pure $ popaque $ pcon PUnit
+      )
+      (validateSum @(n + 1) @xs constr fields)
+
+instance {-# OVERLAPPING #-} SumValidation n '[] where
+  validateSum _ _ = ptraceError "reached end of sum while still not having found the constructor"
+
+----------------------- other utility functions -----------------------------------------
+
+{- | if there is an instance to recover something that is unwrapped from something that is
+ unwrapped, then there is also the possibility to recover the whole thing but wrapped
+-}
+instance
+  ( PTryFrom a b
+  , PIsData a
+  , PIsData b
+  ) =>
+  PTryFrom (PAsData a) (PAsData b)
```

Comment:
I think it's better to encourage users to do these conversions on `PData` only. The only time I see a `PAsData a` -> `PAsData b` conversion being relevant is if the `a` is `PData`, in which case, you should probably just use `pforgetData`. Am I missing something?